roomserver/api/wrapper.go

```diff
@@ -47,34 +47,66 @@ func SendEventWithState(
 	ctx context.Context, rsAPI RoomserverInternalAPI, state *gomatrixserverlib.RespState,
 	event gomatrixserverlib.HeaderedEvent, haveEventIDs map[string]bool,
 ) error {
-	outliers, err := state.Events()
+	isCurrentState := map[string]struct{}{}
+	for _, se := range state.StateEvents {
+		isCurrentState[se.EventID()] = struct{}{}
+	}
+
+	authAndStateEvents, err := state.Events()
 	if err != nil {
 		return err
 	}
 
 	var ires []InputRoomEvent
-	for _, outlier := range outliers {
-		if haveEventIDs[outlier.EventID()] {
+	var stateIDs []string
+
+	for _, authOrStateEvent := range authAndStateEvents {
+		if authOrStateEvent.StateKey() == nil {
 			continue
 		}
-		ires = append(ires, InputRoomEvent{
-			Kind:         KindOutlier,
-			Event:        outlier.Headered(event.RoomVersion),
-			AuthEventIDs: outlier.AuthEventIDs(),
-		})
-	}
+		if haveEventIDs[authOrStateEvent.EventID()] {
+			continue
+		}
+
+		// Work out whether we should store the event as an outlier
+		// or if we should send it as a rewrite event.
+		storeAsOutlier := false
+		if authOrStateEvent.Type() == event.Type() && *authOrStateEvent.StateKey() == *event.StateKey() {
+			// The event is the same as the event that we are looking
+			// to send as a new event.
+			storeAsOutlier = true
+		} else if _, ok := isCurrentState[authOrStateEvent.EventID()]; !ok {
+			// The event doesn't belong to the current state but is
+			// actually an auth event.
+			storeAsOutlier = true
+		}
 
-	stateEventIDs := make([]string, len(state.StateEvents))
-	for i := range state.StateEvents {
-		stateEventIDs[i] = state.StateEvents[i].EventID()
+		if storeAsOutlier {
+			ires = append(ires, InputRoomEvent{
+				Kind:         KindOutlier,
+				Event:        authOrStateEvent.Headered(event.RoomVersion),
+				AuthEventIDs: authOrStateEvent.AuthEventIDs(),
+			})
+		} else {
+			ires = append(ires, InputRoomEvent{
+				Kind:          KindRewrite,
+				Event:         authOrStateEvent.Headered(event.RoomVersion),
+				AuthEventIDs:  authOrStateEvent.AuthEventIDs(),
+				HasState:      true,
+				StateEventIDs: stateIDs,
+			})
+			stateIDs = append(stateIDs, authOrStateEvent.EventID())
```

Comment:
The goal here is to generate:
1. A set of rewrite events, which will form the new state
2. Possibly some outliers, which are auth events but don't form part of the new room state
3. A new event, which is the input event we're dealing with

Each time we process a rewrite event, we want it to have a state snapshot that refers to all of the rewrite events that came before it. Then it gets added to the list so that it appears in the state snapshot of the next rewrite event.

Then by the time we get to the new event at the end, we have a full state comprising all of the rewrite events.

Reply:
This definitely needs clarification in the code then as I thought this was a bug!

Reply:
I'll comment it out better :-)